In [2]:
import pandas as pd

ds = pd.read_csv('../data/train.csv')
ds

id                                             review  sentiment
0          0  Action & Adventure.Billie Clark is twenty year...          1
1          1  I found this early talkie difficult to watch a...          0
2          2  It was Libby talking to Desmond in the flashba...          1
3          3  We fans of Ed Wood tend to be an obsessive bun...          1
4          4  Hmm, Hip Hop music to a period western. Modern...          0
...      ...                                                ...        ...
39577  39577  This story had a different angle that intrigue...          0
39578  39578  I thought the movie (especially the plot) need...          0
39579  39579  The title of this film nearly put me off watch...          1
39580  39580  Mmm, doesn't a big stack of pancakes sound goo...          0
39581  39581  I enjoyed this film. I thought it was an excel...          1

[39582 rows x 3 columns]

In [3]:
from sklearn.model_selection import train_test_split

ds_sample, _ = train_test_split(ds, train_size=6000)
ds_sample

id                                             review  sentiment
29981  29981  I watched this movie at the first showing avai...          1
34765  34765  A clever, undeniably entertaining romp starrin...          1
32373  32373  In 1984, The Karate Kid had some charm to it, ...          0
5759    5759  Along with "Brothers & Sisters", "Six Degrees"...          1
25572  25572  I was watching this movie at one of my usual t...          1
...      ...                                                ...        ...
35844  35844  I watched this film with my family over a long...          1
23196  23196  This is a must see for independant movie fans,...          1
38287  38287  This is a very unusual film in that the star w...          1
15595  15595  I have watched this movie a few times and neve...          0
20931  20931  Every Sunday is an eleven minute short subject...          1

[6000 rows x 3 columns]

In [4]:
txt = ds_sample.review.iloc[1]
txt

'A clever, undeniably entertaining romp starring Peter Ustinov as a career embezzler with his sights set on a US conglomerate in London. He\'s abetted by foxy (and deceptively sharp) Maggie Smith and threatened with exposure by jealous company man Bob Newhart. This is a heist film with a lot of brains. Ustinov is exceptional as is Smith and Newhart is quite funny. The real surprise is Karl Malden, as the pill popping "executive vice-president." Malden has seldom been so loose. Cesar Romero and Robert Morley pop up in some funny cameos. The excellent music score is by Laurie Johnson. Eric Till\'s direction is brisk and the script by Ustinov and Ira Wallach is first-rate & very smart. A swinging good time!'

In [5]:
import nltk

nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/nikita_gordia/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [6]:
nltk.word_tokenize(txt)

['A',
 'clever',
 ',',
 'undeniably',
 'entertaining',
 'romp',
 'starring',
 'Peter',
 'Ustinov',
 'as',
 'a',
 'career',
 'embezzler',
 'with',
 'his',
 'sights',
 'set',
 'on',
 'a',
 'US',
 'conglomerate',
 'in',
 'London',
 '.',
 'He',
 "'s",
 'abetted',
 'by',
 'foxy',
 '(',
 'and',
 'deceptively',
 'sharp',
 ')',
 'Maggie',
 'Smith',
 'and',
 'threatened',
 'with',
 'exposure',
 'by',
 'jealous',
 'company',
 'man',
 'Bob',
 'Newhart',
 '.',
 'This',
 'is',
 'a',
 'heist',
 'film',
 'with',
 'a',
 'lot',
 'of',
 'brains',
 '.',
 'Ustinov',
 'is',
 'exceptional',
 'as',
 'is',
 'Smith',
 'and',
 'Newhart',
 'is',
 'quite',
 'funny',
 '.',
 'The',
 'real',
 'surprise',
 'is',
 'Karl',
 'Malden',
 ',',
 'as',
 'the',
 'pill',
 'popping',
 '``',
 'executive',
 'vice-president',
 '.',
 "''",
 'Malden',
 'has',
 'seldom',
 'been',
 'so',
 'loose',
 '.',
 'Cesar',
 'Romero',
 'and',
 'Robert',
 'Morley',
 'pop',
 'up',
 'in',
 'some',
 'funny',
 'cameos',
 '.',
 'The',
 'excellent',
 '

In [7]:
import string

def remove_punctuation(str):
    return "".join([c for c in str if c not in string.punctuation])

def remove_numbers(str):
    return "".join([c for c in str if c not in string.digits])

txt1 = remove_punctuation(txt)
txt1

'A clever undeniably entertaining romp starring Peter Ustinov as a career embezzler with his sights set on a US conglomerate in London Hes abetted by foxy and deceptively sharp Maggie Smith and threatened with exposure by jealous company man Bob Newhart This is a heist film with a lot of brains Ustinov is exceptional as is Smith and Newhart is quite funny The real surprise is Karl Malden as the pill popping executive vicepresident Malden has seldom been so loose Cesar Romero and Robert Morley pop up in some funny cameos The excellent music score is by Laurie Johnson Eric Tills direction is brisk and the script by Ustinov and Ira Wallach is firstrate  very smart A swinging good time'

In [8]:
from nltk.corpus import stopwords

def clean_stopwords(words):
    stop = stopwords.words('english')
    return [w for w in words if w not in stop]

def tokenize(txt):
    return nltk.word_tokenize(txt)

def lower(txt):
    return txt.lower()

def replace_numbers(words):
    return [("somenumber" if any(ch.isdigit() for ch in w) else w) for w in words]

txt2 = replace_numbers(clean_stopwords(tokenize(lower(txt1))))
txt2

['clever',
 'undeniably',
 'entertaining',
 'romp',
 'starring',
 'peter',
 'ustinov',
 'career',
 'embezzler',
 'sights',
 'set',
 'us',
 'conglomerate',
 'london',
 'hes',
 'abetted',
 'foxy',
 'deceptively',
 'sharp',
 'maggie',
 'smith',
 'threatened',
 'exposure',
 'jealous',
 'company',
 'man',
 'bob',
 'newhart',
 'heist',
 'film',
 'lot',
 'brains',
 'ustinov',
 'exceptional',
 'smith',
 'newhart',
 'quite',
 'funny',
 'real',
 'surprise',
 'karl',
 'malden',
 'pill',
 'popping',
 'executive',
 'vicepresident',
 'malden',
 'seldom',
 'loose',
 'cesar',
 'romero',
 'robert',
 'morley',
 'pop',
 'funny',
 'cameos',
 'excellent',
 'music',
 'score',
 'laurie',
 'johnson',
 'eric',
 'tills',
 'direction',
 'brisk',
 'script',
 'ustinov',
 'ira',
 'wallach',
 'firstrate',
 'smart',
 'swinging',
 'good',
 'time']

In [9]:
def lem(words):
    lemmer = nltk.WordNetLemmatizer()
    return [lemmer.lemmatize(word) for word in words]

def stem(words):
    stemmer = nltk.stem.snowball.EnglishStemmer()
    return [stemmer.stem(word) for word in words]

lem(txt2)

['clever',
 'undeniably',
 'entertaining',
 'romp',
 'starring',
 'peter',
 'ustinov',
 'career',
 'embezzler',
 'sight',
 'set',
 'u',
 'conglomerate',
 'london',
 'he',
 'abetted',
 'foxy',
 'deceptively',
 'sharp',
 'maggie',
 'smith',
 'threatened',
 'exposure',
 'jealous',
 'company',
 'man',
 'bob',
 'newhart',
 'heist',
 'film',
 'lot',
 'brain',
 'ustinov',
 'exceptional',
 'smith',
 'newhart',
 'quite',
 'funny',
 'real',
 'surprise',
 'karl',
 'malden',
 'pill',
 'popping',
 'executive',
 'vicepresident',
 'malden',
 'seldom',
 'loose',
 'cesar',
 'romero',
 'robert',
 'morley',
 'pop',
 'funny',
 'cameo',
 'excellent',
 'music',
 'score',
 'laurie',
 'johnson',
 'eric',
 'till',
 'direction',
 'brisk',
 'script',
 'ustinov',
 'ira',
 'wallach',
 'firstrate',
 'smart',
 'swinging',
 'good',
 'time']

In [9]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import cross_val_score
from sklearn.feature_extraction.text import CountVectorizer

def eval_data(*txt_fn, dataset=ds_sample, max_features=None):
    def prep_txt(txt):
        res = txt
        for fn in txt_fn:
            res = fn(res)
        return res

    count = CountVectorizer(analyzer=prep_txt, max_features=max_features)
    X = count.fit_transform(dataset.review)
    model = MultinomialNB()
    print(cross_val_score(model, X, dataset.sentiment).mean())

eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    stem,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8343333333333334


In [116]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.834


In [117]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    stem,
    clean_stopwords
)

0.8323333333333334


In [118]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    stem,
    clean_stopwords
)

0.8331666666666667


In [119]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    lem,
    stem
)

0.8233333333333335


In [120]:
eval_data(
    lower,
    tokenize,
    clean_stopwords,
    stem,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.828


In [121]:
eval_data(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    stem,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8360000000000001


In [122]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    stem,
    lem
)

0.8333333333333333


### No stem
Best: 0.834

eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

In [123]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem
)

0.8338333333333333


In [124]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    lem
)

0.8258333333333333


In [125]:
eval_data(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8370000000000001


In [126]:
eval_data(
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8283333333333334


In [127]:
eval_data(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem
)

0.8338333333333333


### No remove numbers

0.8370000000000001

eval_data(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

In [128]:
eval_data(
    remove_punctuation,
    lower,
    tokenize,
    replace_numbers,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8335000000000001


### Test max features

In [131]:
eval_data(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=2000
)

0.8401666666666667


In [136]:
eval_data(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=2200
)

0.8388333333333333


In [137]:
eval_data(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=1700
)

0.8400000000000001


### Test n-grams

In [10]:
class LemmaTokenizer:
    def __init__(self):
        self.wnl = nltk.WordNetLemmatizer()
    def __call__(self, doc):
        return clean_stopwords(lem(clean_stopwords(tokenize(doc))))

def preprocessor(input):
    return lower(remove_punctuation(input))

# count = CountVectorizer(tokenizer=LemmaTokenizer(),
#                         preprocessor=preprocessor,
#                         max_features=4500,
#                         ngram_range=(1, 3))
# X = count.fit_transform(ds_sample.review)
# model = MultinomialNB()
# print(cross_val_score(model, X, ds_sample.sentiment).mean())

In [155]:
count = CountVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        max_features=4800,
                        ngram_range=(1, 3))
X = count.fit_transform(ds_sample.review)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

0.8471666666666666


In [160]:
count = CountVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        max_features=5500,
                        ngram_range=(1, 3))
X = count.fit_transform(ds_sample.review)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())


0.8468333333333333


In [161]:

count = CountVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        max_features=5300,
                        ngram_range=(1, 3))
X = count.fit_transform(ds_sample.review)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

0.8465


In [23]:
count = CountVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        max_features=5400,
                        ngram_range=(1, 3))
X = count.fit_transform(ds_sample.review)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

0.8526666666666667


In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer

def eval_data_tfidf(*txt_fn, dataset=ds_sample, max_features=None):
    def prep_txt(txt):
        res = txt
        for fn in txt_fn:
            res = fn(res)
        return res

    count = TfidfVectorizer(analyzer=prep_txt, max_features=max_features)
    X = count.fit_transform(dataset.review)
    model = MultinomialNB()
    print(cross_val_score(model, X, dataset.sentiment).mean())

# eval_data_tfidf(
#     remove_punctuation,
#     remove_numbers,
#     lower,
#     tokenize,
#     clean_stopwords,
#     stem,
#     clean_stopwords,
#     lem,
#     clean_stopwords
# )

In [25]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8486666666666667


In [26]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    stem,
    clean_stopwords
)

0.845


In [40]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    stem,
    clean_stopwords
)

0.845


In [27]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    lem,
    stem
)

0.8351666666666666


In [28]:
eval_data_tfidf(
    lower,
    tokenize,
    clean_stopwords,
    stem,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8431666666666666


In [29]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    stem,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8448333333333334


In [30]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    stem,
    lem
)

0.844


### No stem
Best: 0.8486666666666667

eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

In [31]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem
)

0.8486666666666667


In [32]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    lem
)

0.8393333333333335


In [41]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.851


In [34]:
eval_data_tfidf(
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8440000000000001


In [35]:
eval_data_tfidf(
    remove_punctuation,
    remove_numbers,
    lower,
    tokenize,
    clean_stopwords,
    lem
)

0.8486666666666667


### No remove numbers

0.851

eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords
)

In [36]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    replace_numbers,
    clean_stopwords,
    lem,
    clean_stopwords
)

0.8485000000000001


### Test max features

In [42]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=1000
)

0.8358333333333334


In [43]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=2000
)

0.8445


In [44]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=3000
)

0.8478333333333333


In [45]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=4000
)

0.8463333333333335


In [46]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=5000
)

0.8476666666666667


In [47]:
eval_data_tfidf(
    remove_punctuation,
    lower,
    tokenize,
    clean_stopwords,
    lem,
    clean_stopwords,
    max_features=6000
)

0.8468333333333333


### Test n-grams for TDIDF

In [48]:

count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        ngram_range=(1, 2))
X = count.fit_transform(ds_sample.review)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

0.8641666666666665


In [50]:
count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        ngram_range=(1, 3))
X = count.fit_transform(ds_sample.review)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

0.8651666666666668


In [51]:
count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        ngram_range=(1, 4))
X = count.fit_transform(ds_sample.review)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

0.865


In [35]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.model_selection import cross_val_score

count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        min_df=20,
                        ngram_range=(1, 2))
X = count.fit_transform(ds_sample.review)
print(X.shape)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

(6000, 5408)
0.853


In [38]:
count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        min_df=20,
                        ngram_range=(1, 5))
X = count.fit_transform(ds_sample.review)
print(X.shape)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

(6000, 5442)
0.853


In [37]:
count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        min_df=20,
                        ngram_range=(1, 4))
X = count.fit_transform(ds_sample.review)
print(X.shape)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

(6000, 5441)
0.8531666666666666


In [34]:
count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        min_df=40)
X = count.fit_transform(ds_sample.review)
print(X.shape)
model = MultinomialNB()
print(cross_val_score(model, X, ds_sample.sentiment).mean())

(6000, 2344)
0.8466666666666667


### Save results

In [42]:
from sklearn.feature_extraction.text import CountVectorizer
import numpy as np

count = CountVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        max_features=5400,
                        ngram_range=(1, 3))

np.savez_compressed('../data/dev/feed_count',
        count.fit_transform(ds.review).toarray())

In [24]:
import numpy as np

tfidf = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        max_features=20000,
                        ngram_range=(1, 3))

np.save('../data/dev/feed_tfidf.npy',
        tfidf.fit_transform(ds.review).toarray())

In [40]:
import numpy as np

count = TfidfVectorizer(tokenizer=LemmaTokenizer(),
                        preprocessor=preprocessor,
                        min_df=20,
                        ngram_range=(1, 4))

print('ready to save')
array = tfidf.fit_transform(ds.review).toarray()
print(array.shape)

np.savez_compressed('../data/dev/feed_tfidf_lite', array)

ready to save
(39582, 33559)
